src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -46,102 +38,125 @@ public async Task InvokeAsync(HttpContext context)
+                            await WriteErrorResponseAsync(context, writer, cancellationToken, "Body text could not be parsed. Body text should start with '{' for normal graphql query or with '[' for batched query.").ConfigureAwait(false);
                             return;
                         }
+                        bodyGQLRequest = deserializationResult.Single;
+                        bodyGQLBatchRequest = deserializationResult.Batch;
                         break;
 
-                    case GraphQLContentType:
-                        gqlRequest.Query = await ReadAsStringAsync(httpRequest.Body).ConfigureAwait(false);
+                    case MediaType.GraphQL:
+                        bodyGQLRequest = await DeserializeFromGraphBodyAsync(httpRequest.Body).ConfigureAwait(false);
                         break;
 
-                    case FormUrlEncodedContentType:
+                    case MediaType.Form:
                         var formCollection = await httpRequest.ReadFormAsync().ConfigureAwait(false);
-                        ExtractGraphQLRequestFromPostBody(formCollection, gqlRequest);
+                        bodyGQLRequest = _deserializer.DeserializeFromFormBody(formCollection);
                         break;
 
                     default:
-                        await WriteBadRequestResponseAsync(context, writer, $"Invalid 'Content-Type' header: non-supported media type. Must be of '{JsonContentType}', '{GraphQLContentType}', or '{FormUrlEncodedContentType}'. See: http://graphql.org/learn/serving-over-http/.").ConfigureAwait(false);
+                        await WriteErrorResponseAsync(context, writer, cancellationToken, $"Invalid 'Content-Type' header: non-supported media type. Must be of '{MediaType.Json}' or '{MediaType.GraphQL}'. {_docsUrl}").ConfigureAwait(false);
                         return;
                 }
             }
 
-            IDictionary<string, object> userContext = null;
-            var userContextBuilder = context.RequestServices.GetService<IUserContextBuilder>();
-
-            if (userContextBuilder != null)
+            // If we don't have a batch request, parse the URL too to determine the actual request to run
+            // Querystring params take priority
+            GraphQLRequest gqlRequest = null;
+            if (bodyGQLBatchRequest == null)
             {
-                userContext = await userContextBuilder.BuildUserContext(context).ConfigureAwait(false);
+                // Parse URL
+                var urlGQLRequest = _deserializer.DeserializeFromQueryString(httpRequest.Query);
+
+                gqlRequest = new GraphQLRequest
+                {
+                    Query = urlGQLRequest.Query ?? bodyGQLRequest?.Query,
+                    Inputs = urlGQLRequest.Inputs ?? bodyGQLRequest?.Inputs,
+                    OperationName = urlGQLRequest.OperationName ?? bodyGQLRequest?.OperationName
```

Comment:
It’s good that support for this behavior has appeared, now one can use `application/graphql` even with parametrized queries and pass variables in the request query. 🥇

Reply:
Yea when I added all these other request type handling originally (GET, content types, etc.) I followed the wording on the docs url page, but it's not as clear as just following the express-graphql implementation, which is much more obvious to copy. 

With the new tests and all, we're in a much better place.